### DISCUSSION TITLE
posts/golang/graceful-shutdown-in-go/

### DISCUSSION THREAD

**Discussion by giscus** [josestg/josestg#3] [post-comments] [answered]
# posts/golang/graceful-shutdown-in-go/

Learn how graceful shutdown ensures your requests stay safe during service updates by following the 12-factor principles

https://www.josestg.com/posts/golang/graceful-shutdown-in-go/

**Comment by ariexx** [comment]
Kalo ada koneksi database harus di close juga kah atau servernya doang di close ?

**Reply by josestg** [reply, marked as answer]
simplenya harus juga ya. Normalnya kita nulis code gini di Go, jadi ketika server nya graceful shutdown otomatis db nya juga close ya. Karna ada `defer`.

```go
func run() error {
     db, err :=  sql.Open(....)
      if err != nil {
           return err
      }
     defer db.Close()

     s := http.Server{Addr: ":8080"}
     gs := wrapWithGracefulShutdown(s)
     return gs.ListenAndServe()
}
```
